system/datatypes/mailinglist.php

```diff
@@ -193,19 +198,20 @@ public function indexAction() {
 
 				if ($addAddress) {
 					// add email with status pending
-					$this->xml->lockAndReload();
 					$address = $this->xml->createElement('address');
 					$address->generateId();
 					$address->setAttribute('email', $email);
 					$address->setAttribute('status', self::ADDRESS_STATUS_PENDING);
 					$address->setAttribute('confirm-code', $code);
 					$this->getDoc()->getOrCreate('addresses')->append($address);
 					$this->xml->saveAndUnlock();
+				} else {
+					$this->xml->unlock();
 				}
 
 				if ($sendEmail) {
 					// send email so that action can be confirmed
-					$link = $this->generateFullPath(sprintf('%s/confirm?email=%s&code=%s', $this->pagename, $email, $code));
+					$link = $this->generateFullPath(sprintf('%s/confirm?email=%s&code=%s', $this->pagename, urlencode($email), $code));
```

Comment:
For consistency, I would urlencode all three parts. What if the pagename contains an `&`, or we switch to an code-generation method that outputs `%` or `&` or whatever? Just that something is not (directly) user-controlled, or previously verified / sanitized, is (IMHO) no reason to not encode it anyway.